<div class="alert alert-info">
Привет, Александра! Меня зовут Светлана Чих и я буду проверять твой проект. Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе.

<div class="alert alert-success">
<b>👍 Успех:</b> Зелёным цветом отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах.
</div>
<div class="alert alert-warning">
<b>🤔 Рекомендация:</b> Жёлтым цветом выделено то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас (однако это не обязательно).
</div>
<div class="alert alert-danger">
<b>😔 Необходимо исправить:</b> Красным цветом выделены комментарии, без исправления которых, я не смогу принять проект :(
</div>
<div class="alert alert-info">
<b>👂 Совет:</b> Какие-то дополнительные материалы
</div>
Давай работать над проектом в диалоге: если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом.
Мне будет легче отследить изменения, если ты выделишь свои комментарии:
<div class="alert alert-info"> <b>🎓 Комментарий студента:</b> Например, вот так.</div>
Пожалуйста, не перемещай, не изменяй и не удаляй мои комментарии. Всё это поможет выполнить повторную проверку твоего проекта быстрее.
 </div>

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-задачи" data-toc-modified-id="Описание-задачи-1">Описание задачи</a></span></li><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-2">Описание данных</a></span></li><li><span><a href="#Иморт-библиотек-и-подключение-к-БД" data-toc-modified-id="Иморт-библиотек-и-подключение-к-БД-3">Иморт библиотек и подключение к БД</a></span></li><li><span><a href="#Вывод-таблиц" data-toc-modified-id="Вывод-таблиц-4">Вывод таблиц</a></span></li><li><span><a href="#Задание-1" data-toc-modified-id="Задание-1-5">Задание 1</a></span></li><li><span><a href="#Задание-2" data-toc-modified-id="Задание-2-6">Задание 2</a></span></li><li><span><a href="#Задание-3" data-toc-modified-id="Задание-3-7">Задание 3</a></span></li><li><span><a href="#Задание-4" data-toc-modified-id="Задание-4-8">Задание 4</a></span></li><li><span><a href="#Задание-5" data-toc-modified-id="Задание-5-9">Задание 5</a></span></li></ul></div>

# SQL

## Описание задачи

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Иморт библиотек и подключение к БД

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

## Вывод таблиц

In [2]:
#выведем первые 10 строк каждой таблицы
#таблица books

books = "SELECT * FROM books LIMIT 10"
pd.io.sql.read_sql(books, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

In [3]:
#таблица authors

authors = "SELECT * FROM authors LIMIT 10"
pd.io.sql.read_sql(authors, con = engine)

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

In [4]:
#таблица publishers

publishers = "SELECT * FROM publishers LIMIT 10"
pd.io.sql.read_sql(publishers, con = engine)

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

In [5]:
#таблица ratings

ratings = "SELECT * FROM ratings LIMIT 10"
pd.io.sql.read_sql(ratings, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

In [6]:
#таблица reviews

reviews = "SELECT * FROM reviews LIMIT 10"
pd.io.sql.read_sql(reviews, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

<div class="alert alert-success">
<b>👍 Успех:</b> Есть описание задачи, импортированы все нужные библиотеки, создано подключение к БД, просмотрены таблицы
</div>

## Задание 1

Посчитайте, сколько книг вышло после 1 января 2000 года

In [7]:
query1 = "SELECT COUNT(book_id) FROM books WHERE publication_date > '2000-01-01'"

pd.io.sql.read_sql(query1, con = engine)

count
0    819

После 1 января 2000 года вышло 819 книг

<div class="alert alert-success">
<b>👍 Успех:</b> Все верно!
</div>

## Задание 2

Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [8]:
query2 =  "SELECT a.author, b.title, COUNT(DISTINCT(rv.review_id)), AVG(rt.rating) FROM books b LEFT JOIN reviews rv ON b.book_id = rv.book_id LEFT JOIN ratings rt ON b.book_id = rt.book_id LEFT JOIN authors a ON b.author_id = a.author_id GROUP BY b.book_id, a.author, b.title ORDER BY COUNT(rv.review_id) DESC"
pd.io.sql.read_sql(query2, con = engine)

author  \
0                                      Stephenie Meyer   
1                                       J.R.R. Tolkien   
2                                        J.D. Salinger   
3                           J.K. Rowling/Mary GrandPré   
4                           J.K. Rowling/Mary GrandPré   
..                                                 ...   
995                 Edgar Allan Poe/Benjamin F. Fisher   
996                      Leonardo da Vinci/H. Anna Suh   
997                                  Dr. Seuss/Various   
998  Faye Perozich/Anne Rice/John Bolton/Daerick Gröss   
999                      Kimon Nicolaides/Mamie Harmon   

                                                 title  count       avg  
0                              Twilight (Twilight  #1)      7  3.662500  
1                  The Hobbit  or There and Back Again      6  4.125000  
2                               The Catcher in the Rye      6  3.825581  
3    Harry Potter and the Prisoner of Azkaban (Harr...      6  4.414634  
4    Harry Potter and the Chamber of Secrets (Harry...      6  4.287500  
..                                                 ...    ...       ...  
995                          Essential Tales and Poems      0  4.000000  
996                               Leonardo's Notebooks      0  4.000000  
997   The Cat in the Hat and Other Dr. Seuss Favorites      0  5.000000  
998    Anne Rice's The Vampire Lestat: A Graphic Novel      0  3.666667  
999                            The Natural Way to Draw      0  3.000000  

[1000 rows x 4 columns]

Отсортировали таблицу по убыванию количества обзоров, чтобы понять, какие книги собрали больше всего мнений. 

Получили пятерку лидеров - самых популярных книг у читателей:
 1) Сумерки 1 часть
 
 2) Хоббит
 
 3) Над пропастью во ржи
 
 4) Гарри Поттер и узник Азкабана
 
 5) Гарри Поттер и тайная комната
 
Не сортировали по столбцу с рейтингом, т.к., на мой взгляд, кол-во отзывов лучше показывает отношение читателей к произведению.

<div class="alert alert-success">
<b>👍 Успех:</b> Все верно!
</div>

## Задание 3

Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [9]:
query3 = "SELECT p.publisher, COUNT(b.book_id) FROM publishers p LEFT JOIN books b ON p.publisher_id = b.publisher_id WHERE b.num_pages > 50 GROUP BY p.publisher ORDER BY COUNT(b.book_id) DESC"
pd.io.sql.read_sql(query3, con = engine)

publisher  count
0                                        Penguin Books     42
1                                              Vintage     31
2                             Grand Central Publishing     25
3                                     Penguin Classics     24
4                                     Ballantine Books     19
..                                                 ...    ...
329                                         Turtleback      1
330  Atheneum Books for Young Readers: Richard Jack...      1
331                                     Penguin Signet      1
332                                    Victor Gollancz      1
333                      Harvard Business Review Press      1

[334 rows x 2 columns]

Издательство с наибольшим количеством книг, содержащих более 50 стр, - **Penguin Books**

<div class="alert alert-success">
<b>👍 Успех:</b> Все верно!
</div>

## Задание 4

Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [10]:
query4 = "SELECT a.author, AVG(rt.rating) FROM authors a LEFT JOIN books b ON a.author_id = b.author_id LEFT JOIN ratings rt ON b.book_id = rt.book_id WHERE b.book_id IN (SELECT b.book_id FROM books b LEFT JOIN ratings rt ON b.book_id = rt.book_id GROUP BY b.book_id HAVING COUNT(rt.rating) >= 50ORDER BY COUNT(rt.rating) DESC) GROUP BY a.author ORDER BY AVG(rt.rating) DESC"

pd.io.sql.read_sql(query4, con = engine)

author       avg
0                          J.K. Rowling/Mary GrandPré  4.287097
1                   Markus Zusak/Cao Xuân Việt Khương  4.264151
2                                      J.R.R. Tolkien  4.246914
3                                   Louisa May Alcott  4.192308
4                                        Rick Riordan  4.080645
5                                     William Golding  3.901408
6                                       J.D. Salinger  3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce  3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...  3.787879
9                                          Lois Lowry  3.750000
10                                          Dan Brown  3.741259
11            George Orwell/Boris Grabnar/Peter Škerl  3.729730
12                                    Stephenie Meyer  3.662500
13                                     John Steinbeck  3.622951

У Джоан Роулинг самая высокая средняя оценка книг (по книгам, имеющим 50 и более оценок)

<div class="alert alert-success">
<b>👍 Успех:</b> Все верно!
</div>

## Задание 5

Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [11]:
query5 = " WITH i AS (SELECT username, COUNT(review_id) FROM reviews WHERE username IN (SELECT rt.username FROM ratings rt GROUP BY rt.username HAVING COUNT(rt.rating) > 50) GROUP BY username) SELECT AVG(i.count) FROM i"

pd.io.sql.read_sql(query5, con = engine)

avg
0  24.333333

Пользователями, которые поставили больше 50 оценок, было написано в среднем 24 обзора

<div class="alert alert-success">
<b>👍 Успех:</b> Все верно!
</div>

<div class="alert alert-success">
<b>👍 Успех:</b> Ты справилась с работой и со всем курсом, поздравляю!
</div>